edk2toollib/windows/policy/firmware_policy.py

```diff
@@ -0,0 +1,563 @@
+        localArray = bytearray()
+        self.SubKeyName.Serialize(valueOut=localArray)
+        valueOut += localArray
+
+        self.OffsetToValueName = self.OffsetToSubKeyName + len(localArray)
+        localArray = bytearray()
+        self.ValueName.Serialize(valueOut=localArray)
+        valueOut += localArray
+
+        self.OffsetToValue = self.OffsetToValueName + len(localArray)
+        localArray = bytearray()
+        self.Value.Serialize(valueOut=localArray)
+        valueOut += localArray
+
+        ruleOut += struct.pack(self.StructFormat, self.RootKey, self.OffsetToSubKeyName,
+                               self.OffsetToValueName, self.OffsetToValue)
+
+
+##
+# Class for storing, serializing, deserializing, & printing PolicyValue Types
+##
+class PolicyValueType():
+    StructFormat = '<H'
+    StructSize = struct.calcsize(StructFormat)
+
+    POLICY_VALUE_TYPE_STRING = 0
+    POLICY_VALUE_TYPE_BOOL = 1
+    POLICY_VALUE_TYPE_DWORD = 2
+    POLICY_VALUE_TYPE_DWORD_RANGE = 3
+    POLICY_VALUE_TYPE_DWORD_CHOICE = 4
+    POLICY_VALUE_TYPE_QWORD = 5
+    POLICY_VALUE_TYPE_QWORD_RANGE = 6
+    POLICY_VALUE_TYPE_QWORD_CHOICE = 7
+    POLICY_VALUE_TYPE_OPTION = 8
+    POLICY_VALUE_TYPE_MULTI_STRING = 9
+    POLICY_VALUE_TYPE_BINARY = 10
+
+    SupportedValueTypes = {POLICY_VALUE_TYPE_DWORD,
+                           POLICY_VALUE_TYPE_QWORD,
+                           POLICY_VALUE_TYPE_STRING}
+
+    def __init__(self, Type):
+        if Type not in self.SupportedValueTypes:
+            ErrorMessage = ('Unsupported ValueType: %x' % Type)
+            print(ErrorMessage)
+            if STRICT:
+                raise Exception(ErrorMessage)
+
+        self.vt = Type
+
+    # if offset is not specified, fs is assumed pointing at the PolicyValueType struct
+    @classmethod
+    def FromFileStream(cls, fs, offset=None):
+        if offset:
+            fs.seek(offset)
+        else:
+            offset = fs.tell()
+        valueType = struct.unpack('<H', fs.read(2))[0]
+        return cls(valueType)
+
+    def Print(self, prefix=''):
+        print('%s%s%s' % (prefix, 'ValueType: ', self.vt))
+
+    def Serialize(self, valueOut):
+        valueOut += struct.pack(self.StructFormat, self.vt)
+
+    def Get(self):
+        return self.vt
+
+
+##
+# Class for storing, serializing, deserializing, & printing PolicyString Types
+# NOTE: This type is used both in the keys as SubKeyName and ValueName and it
+#       can be a value.  When used as a value, a NULL may follow the string, but
+#       the NULL will not be included in the string size
+##
+class PolicyString():
+    #
+    # 16-bit, little endian, size of the string in _bytes_
+    # followed by a UTF-16LE string, no NULL terminator
+    #
+    # Example of "PlatformID"
+    # \x14\x00P\x00l\x00a\x00t\x00f\x00o\x00r\x00m\x00I\x00D\x00
+    #
+    # The trailing \x00 is not a NULL, it is UTF-16LE encoding of "D"
+
+    def __init__(self, fs=None, offset=None, String=None):
```

Comment:
why offset.  It looks like it is optional but you never seek back....so why not make the caller seek to the right point to start and the caller fix the fs pointer on return.

Reply:
It pushes the seek down into the worker classes so that the high level code is more readable IMHO.
A 3 line seek/call/restore can become a 1 line call with offset parameter.  Because the value table is not guaranteed sequential, every value requires 3 seeks.

Reply:
so does the caller do a seek after return to bring FS back to expected location? 

Is it fair to say fs on entry should always point to head of value table?  Should FS just be a value table parameter instead of file stream.  Meaning should the outer object parse the value table as a bye array and then pass that to the inner objects?  This way you would also contain your seek length within the value table

Reply:
addressed in the 10th commit 29d5833845f3e9c31652f9b740786fdef65285df